[Xcode] Fix size increase due to multiple PGO profile support

emw-apple opened the pull request:
#### 99e4e950c68dd8b4fb780155b63caa3e0d70b41c
<pre>
[Xcode] Fix size increase due to multiple PGO profile support
<a href="https://bugs.webkit.org/show_bug.cgi?id=253870">https://bugs.webkit.org/show_bug.cgi?id=253870</a>
rdar://106673505

Reviewed by Alexey Proskuryakov.

The target triple selection logic in copy-profiling-data is moving to
WebKitAdditions. Change the script on the open-source side to simply
extract the profile that WebKitAdditions has built with.

Rather than revert to the old configuration logic where we built our own
OTHER_CFLAGS of PGO switches, use the settings that have been built-in to
Xcode for years. AFAIK, we didn&apos;t know about these settings during the
initial PGO work. As a result of this change, PGO profdata are correctly
tracked as dependencies of the compile tasks that use them. We also save
a bit of xcconfig logic.

Additionally, adjust some use of GCC_PREPROCESSOR_DEFINITIONS in a
drive-by cleanup. Feature-list variables like
ENABLE_LLVM_PROFILE_GENERATION or ENABLE_PDFJS, which expand to their
own name, can be used as -D flags like

    GCC_PREPROCESSOR_DEFINITIONS = $(ENABLE_PDFJS)

instead of the more superfluous

    GCC_PREPROCESSOR_DEFINITIONS = GCC_PREPROCESSOR_DEFINITIONS_$(ENABLE_PDFJS)
    GCC_PREPROCESSOR_DEFINITIONS_ENABLE_PDFJS = ENABLE_PDFJS=1

* Configurations/WebKitProjectPaths.xcconfig:
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Scripts/copy-profiling-data.sh:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/MediaFormatReader.xcconfig:
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/Configurations/adattributiond.xcconfig:
* Source/WebKit/Configurations/webpushd.xcconfig:
* Source/WebKit/Configurations/webpushtool.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:

Canonical link: <a href="https://commits.webkit.org/261778@main">https://commits.webkit.org/261778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12c7e4839091d40ae2ac49a30513a5be194dde00

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112751 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1454 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4560 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121295 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116821 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23282 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13105 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5723 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118520 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100539 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105866 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1080 "1 flakes 4 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46312 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/101057 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14246 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1124 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12380 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14944 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102574 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53113 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/32019 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8209 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16791 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110617 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27321 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
```diff
@@ -50,25 +50,26 @@ SECTORDER_FLAGS = $(SECTORDER_FLAGS_$(CONFIGURATION));
+CLANG_USE_OPTIMIZATION_PROFILE_Release_macosx_MACOS_SINCE_1200 = YES;
+CLANG_USE_OPTIMIZATION_PROFILE_Release_iphoneos = YES;
+CLANG_USE_OPTIMIZATION_PROFILE_Production_macosx = $(CLANG_USE_OPTIMIZATION_PROFILE_Production_macosx$(WK_MACOS_1200));
+CLANG_USE_OPTIMIZATION_PROFILE_Production_macosx_MACOS_SINCE_1200 = YES;
```
We don't need `MACOS_SINCE_1200` any more.

aproskuryakov commented on Source/JavaScriptCore/Scripts/copy-profiling-data.sh:
```diff
@@ -1,66 +1,12 @@
-    exit 1
+if [ "${CLANG_USE_OPTIMIZATION_PROFILE}" = YES ]; then
+    compression_tool -v -decode -i "${SCRIPT_INPUT_FILE_0}" -o "${SCRIPT_OUTPUT_FILE_0}" && exit
+    [ "${CONFIGURATION}" = Production ] && exit 1
```
Was this error silent before? Seems like a common enough code path to print a user friendly error.

emw-apple replied:
On Production builds, it'll fail with the error message from `compression_tool` being the last thing printed. On at-desk Release builds, it'll print the `compression_tool` error message, followed by the warning message that it's continuing with stub profdata.

aproskuryakov replied:
Right, this is what I expected too. I think that a more helpful message would be good for production builds.

emw-apple replied:
I see, sounds good!

aproskuryakov commented on Source/JavaScriptCore/Scripts/copy-profiling-data.sh:
```diff
@@ -1,66 +1,12 @@
+    echo "warning: ${SCRIPT_INPUT_FILE_0} failed to extract, falling back to stub profile data"
 fi
 
+# Copy stub data to the output path to satisfy the build system.
```
It would be instructional to explain what exactly "satisfy the build system" means, so that someone looking to rewrite this in the future would know if it's still necessary.

aproskuryakov reviewed (approved):
Nice, always great to see more lines removed than added!

emw-apple committed:
[Xcode] Fix size increase due to multiple PGO profile support
https://bugs.webkit.org/show_bug.cgi?id=253870
rdar://106673505

Reviewed by Alexey Proskuryakov.

The target triple selection logic in copy-profiling-data is moving to
WebKitAdditions. Change the script on the open-source side to simply
extract the profile that WebKitAdditions has built with.

Rather than revert to the old configuration logic where we built our own
OTHER_CFLAGS of PGO switches, use the settings that have been built-in to
Xcode for years. AFAIK, we didn't know about these settings during the
initial PGO work. As a result of this change, PGO profdata are correctly
tracked as dependencies of the compile tasks that use them. We also save
a bit of xcconfig logic.

Additionally, adjust some use of GCC_PREPROCESSOR_DEFINITIONS in a
drive-by cleanup. Feature-list variables like
ENABLE_LLVM_PROFILE_GENERATION or ENABLE_PDFJS, which expand to their
own name, can be used as -D flags like

    GCC_PREPROCESSOR_DEFINITIONS = $(ENABLE_PDFJS)

instead of the more superfluous

    GCC_PREPROCESSOR_DEFINITIONS = GCC_PREPROCESSOR_DEFINITIONS_$(ENABLE_PDFJS)
    GCC_PREPROCESSOR_DEFINITIONS_ENABLE_PDFJS = ENABLE_PDFJS=1

* Configurations/WebKitProjectPaths.xcconfig:
* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Scripts/copy-profiling-data.sh:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/MediaFormatReader.xcconfig:
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/Configurations/adattributiond.xcconfig:
* Source/WebKit/Configurations/webpushd.xcconfig:
* Source/WebKit/Configurations/webpushtool.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:

Canonical link: https://commits.webkit.org/261778@main
